src/credentials/GroupDataProviderImpl.cpp

```diff
@@ -0,0 +1,1388 @@
+{
+    VerifyOrReturnError(mInitialized, nullptr);
+    return mKeysetIterators.CreateObject(*this, fabric_index);
+}
+
+GroupDataProviderImpl::KeysetIteratorImpl::KeysetIteratorImpl(GroupDataProviderImpl & provider, chip::FabricIndex fabric_index) :
+    mProvider(provider), mFabric(fabric_index)
+{
+    FabricData fabric(fabric_index);
+    if (CHIP_NO_ERROR == fabric.Load(provider.mStorage))
+    {
+        mNextId = fabric.first_keyset;
+        mCount  = fabric.keyset_count;
+        mIndex  = 0;
+    }
+}
+
+size_t GroupDataProviderImpl::KeysetIteratorImpl::Count()
+{
+    return mCount;
+}
+
+bool GroupDataProviderImpl::KeysetIteratorImpl::Next(Keyset & item)
+{
+    VerifyOrReturnError(mIndex < mCount, false);
+
+    KeysetData keyset(mFabric, mNextId);
+    VerifyOrReturnError(CHIP_NO_ERROR == keyset.Load(mProvider.mStorage), false);
+
+    mIndex++;
+    mNextId            = keyset.next;
+    item.keyset_id     = keyset.keyset_id;
+    item.policy        = keyset.policy;
+    item.num_keys_used = keyset.num_keys_used;
+    memcpy(item.epoch_keys, keyset.epoch_keys, sizeof(item.epoch_keys));
+    return true;
+}
+
+void GroupDataProviderImpl::KeysetIteratorImpl::Release()
+{
+    mProvider.mKeysetIterators.ReleaseObject(this);
+}
+
+//
+// Fabrics
+//
+
+CHIP_ERROR GroupDataProviderImpl::RemoveFabric(chip::FabricIndex fabric_index)
+{
+    FabricData fabric(fabric_index);
+    VerifyOrReturnError(CHIP_NO_ERROR == fabric.Load(mStorage), CHIP_ERROR_KEY_NOT_FOUND);
+
+    // Remove Group Mappings
+
+    EndpointData endpoint(fabric_index, fabric.first_endpoint);
+    size_t endpoint_count = 0;
+
+    do
+    {
+        if (CHIP_NO_ERROR != endpoint.Load(mStorage))
+        {
+            break;
+        }
+        RemoveAllGroupMappings(fabric_index, endpoint.id);
+        endpoint.id = endpoint.next;
+    } while ((++endpoint_count < fabric.endpoint_count) && (kInvalidEndpointId != endpoint.id));
+
+    // Remove States
+
+    StateListData states;
+
+    // Load state list info
+    states.Load(mStorage);
+
+    StateData state(states.first);
+    size_t state_index = 0;
+
+    while (state_index < states.count)
+    {
+        if (CHIP_NO_ERROR != state.Load(mStorage))
+        {
+            break;
+        }
+        if (state.fabric_index == fabric_index)
+        {
+            // Removing the state shifts down the upper states
+            RemoveGroupState(state_index);
+        }
+        else
+        {
+            state_index++;
+        }
+        state.index = state.next;
+    }
+
+    // Remove Keysets
+
+    KeysetData keyset(fabric_index, fabric.first_keyset);
+    size_t keyset_count = 0;
+
+    // Loop until the desired index
+    while (keyset_count < fabric.keyset_count)
+    {
+        if (CHIP_NO_ERROR != keyset.Load(mStorage))
+        {
+            break;
+        }
+        RemoveKeyset(fabric_index, keyset.keyset_id);
```

Comment:
Why is there no `if (keyset.fabric_index == fabric_index)` ? Won't this try to erase keyset indices that match a different fabric, by calling a given fabric's RemoveKeyset?

Reply:
I thought this was a case of not enough coffee, but upon closer inspection, the condition is not needed here: Each fabric has a separate linked-list, so all the states in this list are to be removed. I have added a comment that reads:
```
    // Loop the keysets associated with the target fabric
    while (keyset_count < fabric.keyset_count)
    ...
```